### PR TITLE
GSoC 2021 custom picture selector

### PR DESCRIPTION
**Pull request by nicolas-raoul**

### CODE REVIEW

**Comment by misaochan** [comment]
Hi @nicolas-raoul and @4D17Y4 , thanks for the PR! :) Is it OK if we postpone the merge to after our WLM release, which should be out in about 1-2 weeks from now? It looks good to me (the failed build I believe is due to the travis.org issue), but we have very limited time to debug the release prior to WLM start, so we are trying to minimize new features in the release.

File: app/src/main/java/fr/free/nrw/commons/data/DBOpenHelper.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
 public class DBOpenHelper  extends SQLiteOpenHelper {
 
     private static final String DATABASE_NAME = "commons.db";
-    private static final int DATABASE_VERSION = 15;
+    private static final int DATABASE_VERSION = 14;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/data/DBOpenHelper.java:16]: Please make it `16`.

File: app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java [hunk 2]
```diff
@@ -66,6 +68,11 @@ public CommonsApplicationModule(Context applicationContext) {
         this.applicationContext = applicationContext;
     }
 
+    @Provides
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/di/CommonsApplicationModule.java:71]: javadoc

File: app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java [hunk 3]
```diff
@@ -51,6 +53,11 @@ private static Intent createGalleryIntent(@NonNull Context context, int type) {
                 .putExtra(Intent.EXTRA_ALLOW_MULTIPLE, configuration(context).allowsMultiplePickingInGallery());
     }
 
+    private static Intent createCustomSelectorIntent(@NonNull Context context, int type) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java:56]: javadoc

File: app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java [hunk 4]
```diff
@@ -197,6 +215,32 @@ private static void onPictureReturnedFromDocuments(Intent data, Activity activit
         }
     }
 
+    private static void onPictureReturnedFromCustomSelector(Intent data, Activity activity, @NonNull FilePicker.Callbacks callbacks) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java:218]: javadoc

File: app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java [hunk 5]
```diff
@@ -197,6 +215,32 @@ private static void onPictureReturnedFromDocuments(Intent data, Activity activit
...
+        try {
+            List<UploadableFile> files = getFilesFromCustomSelector(data, activity);
+            callbacks.onImagesPicked(files, ImageSource.CUSTOM_SELECTOR, restoreType(activity));
+        } catch (Exception e) {
+            e.printStackTrace();
+            callbacks.onImagePickerError(e, ImageSource.CUSTOM_SELECTOR, restoreType(activity));
+        }
+    }
+
+    private static List<UploadableFile> getFilesFromCustomSelector(Intent data, Activity activity) throws  IOException, SecurityException {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/filepicker/FilePicker.java:228]: javadoc

File: app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java [hunk 6]
```diff
@@ -66,8 +69,10 @@ public int getImageQuality() {
         return imageQuality.getValue();
     }
 
+    public Uri getContentUri() { return contentUri; }
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java:72]: javadoc with example value

File: app/src/main/java/fr/free/nrw/commons/filepicker/PickedFiles.java [hunk 7]
```diff
@@ -25,7 +25,7 @@
 import timber.log.Timber;
 
 
-class PickedFiles implements Constants {
+public class PickedFiles implements Constants {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/filepicker/PickedFiles.java:28]: making it public is a good time to add javadoc :-)

**Review by nicolas-raoul — COMMENTED** [review]
javadoc and aesthetic things

**Comment by nicolas-raoul** [comment]
@misaochan Understood! When it is OK to merge, please let us know :-)

File: app/build.gradle [hunk 8]
```diff
@@ -204,7 +209,7 @@ android {
             }
         }
         debug {
-            minifyEnabled true
+            minifyEnabled false
```
Comment (nicolas-raoul) [inline app/build.gradle:212]: Not needed anymore?

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+import fr.free.nrw.commons.customselector.ui.adapter.ImageAdapter
+import fr.free.nrw.commons.di.CommonsDaggerSupportFragment
+import fr.free.nrw.commons.theme.BaseActivity
+import kotlinx.android.synthetic.main.fragment_custom_selector.view.*
+import java.io.File
+import java.io.FileInputStream
+import java.net.URI
+import javax.inject.Inject
+
+class ImageFragment: CommonsDaggerSupportFragment() {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt:29]: javadoc

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+     * getSpanCount for GridViewManager.
+     *
+     * @return spanCount.
+     */
+    private fun getSpanCount(): Int {
+        return 3
+        // todo change span count depending on the device orientation and other factos.
+    }
+
+    override fun onResume() {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt:172]: javadoc

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+    sealed class Result {
+        object TRUE : Result()
+        object FALSE : Result()
+        object INVALID : Result()
+        object NOTFOUND : Result()
+        object ERROR : Result()
+    }
+
+    companion object {
+        const val INVALIDATE_DAY_COUNT: Long = 7
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:270]: javadoc explaining what this number means, with a simple example.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java [hunk 12]
```diff
@@ -23,6 +23,7 @@
     private final String createdTimestampSource;
     private final BehaviorSubject<Integer> imageQuality;
     private boolean hasInvalidLocation;
+    private final Uri contentUri;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java:26]: Would you mind copy/pasting the getContentUri() javadoc to here?

File: app/build.gradle (hunk 8 above)
Comment (4D17Y4) [reply app/build.gradle:212]: Will fix it thanks.

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt (hunk 9 above)
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt:29]: done.

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt (hunk 10 above)
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageFragment.kt:172]: done.

File: app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt (hunk 11 above)
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/customselector/ui/selector/ImageLoader.kt:270]: done.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java (hunk 12 above)
Comment (4D17Y4) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadItem.java:26]: done.

**Comment by misaochan** [comment]
Hi @nicolas-raoul , you can merge this into master when ready, as we have taken out the 3.1-release branch already. :)

**Review by madhurgupta10 — APPROVED** [review]

**Comment by nicolas-raoul** [comment]
Great job Aditya!

**Comment by misaochan** [comment]
Great job @4D17Y4 :)

**Comment by madhurgupta10** [comment]
@4D17Y4 Congratulations 🎉

**Pull request merged**